src/ddmd/cppmangle.d

```diff
@@ -34,225 +34,267 @@ import ddmd.visitor;
+        auto i = find(p);
+        if (i >= 0)
+        {
+            //printf("\tmatch\n");
+            /* Sequence is S_, S0_, .., S9_, SA_, ..., SZ_, S10_, ...
+             */
+            buf.writeByte('S');
+            if (i)
             {
-                //printf("    component[%d] = %s\n", i, components[i] ? components[i].toChars() : null);
-                if (p == components[i])
+                // Write <seq-id> to buf
+                void write_seq_id(size_t i)
                 {
-                    //printf("\tmatch\n");
-                    /* Sequence is S_, S0_, .., S9_, SA_, ..., SZ_, S10_, ...
-                     */
-                    buf.writeByte('S');
-                    if (i)
-                        writeBase36(i - 1);
-                    buf.writeByte('_');
-                    return true;
+                    if (i >= 36)
+                    {
+                        write_seq_id(i / 36);
+                        i %= 36;
+                    }
+                    i += (i < 10) ? '0' : 'A' - 10;
+                    buf.writeByte(cast(char)i);
                 }
+
+                write_seq_id(i - 1);
             }
+            buf.writeByte('_');
+            return true;
+        }
         return false;
     }
 
-    bool exist(RootObject p)
+    /******
+     * See if `p` exists in components[]
+     * Returns:
+     *  index if found, -1 if not
+     */
+    int find(RootObject p)
     {
-        //printf("exist %s\n", p ? p.toChars() : null);
+        //printf("find %p %d %s\n", p, p.dyncast(), p ? p.toChars() : null);
         if (components_on)
-            for (size_t i = 0; i < components.dim; i++)
+        {
+            foreach (i, component; components)
             {
-                if (p == components[i])
-                {
-                    return true;
-                }
+                if (p == component)
+                    return cast(int)i;
             }
-        return false;
+        }
+        return -1;
     }
 
-    void store(RootObject p)
+    /*********************
+     * Append p to components[]
+     */
+    void append(RootObject p)
     {
-        //printf("store %s\n", p ? p.toChars() : "null");
+        //printf("append %p %d %s\n", p, p.dyncast(), p ? p.toChars() : "null");
         if (components_on)
             components.push(p);
     }
 
-    void source_name(Dsymbol s, bool skipname = false)
+    /******************************
+     * Write the mangled representation of the template arguments.
+     * Params:
+     *  ti = the template instance
+     */
+    void template_args(TemplateInstance ti)
     {
-        //printf("source_name(%s)\n", s.toChars());
-        TemplateInstance ti = s.isTemplateInstance();
-        if (ti)
-        {
-            if (!skipname && !substitute(ti.tempdecl))
+        /* <template-args> ::= I <template-arg>+ E
+         */
+        if (!ti)                // could happen if std::basic_string is not a template
```

Comment:
Love this. Everywhere else the code in this PR has the shape:

```d
T some_function(SomeArguments some_arguments)
{
    if (condition)
    {
        ... many, many, MANY lines of code ...
        return result;
    }
    return corner_case;
}
```

which is distinctly unsavory for some obvious and some less obvious reasons.